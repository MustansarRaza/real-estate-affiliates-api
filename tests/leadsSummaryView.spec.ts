import { loadSequelizeFixtures, useRequestFactory } from "./support";
import { ModelFixtureNames } from "./support/data";

describe("GET /leads/summary", () => {
  beforeEach(async () => {
    await loadSequelizeFixtures(
      ModelFixtureNames.AGENTS,
      ModelFixtureNames.USERS,
      ModelFixtureNames.CONTACTS,
      ModelFixtureNames.CATEGORIES,
      ModelFixtureNames.AFFILIATES,
      ModelFixtureNames.AFFILIATE_CONTACTS,
      ModelFixtureNames.POTENTIAL_AFFILIATE_VERIFICATION_STATUS,
      ModelFixtureNames.POTENTIAL_AFFILIATES,
      ModelFixtureNames.CLIENTS,
      ModelFixtureNames.LEAD_STATUSES,
      ModelFixtureNames.LEADS,
      ModelFixtureNames.TASK_TYPES,
      ModelFixtureNames.TASKS
    );
  });

  it("denies access to unauthenticated requests", async () => {
    const { body, header } = await useRequestFactory()
      .get("/leads/summary?fromDate=2020-01-01T00:00:00Z&toDate=2021-04-15T00:00:00Z")
      .expect(403);

    expect(body).toMatchSnapshot();
    expect(header).toMatchSnapshot();
  });

  it("returns leads summary for a verified user", async () => {
    const { body, header } = await useRequestFactory()
      .withVerifiedUser()
      .get("/leads/summary?fromDate=2021-01-01T00:00:00Z&toDate=2021-04-15T00:00:00Z")
      .expect(200);

    expect(body).toMatchSnapshot();
    expect(header).toMatchSnapshot();
  });

  it("returns an empty summary object for an unverified user", async () => {
    const { body } = await useRequestFactory()
      .withUnverifiedUser()
      .get("/leads/summary?fromDate=2021-01-01T00:00:00Z&toDate=2021-04-15T00:00:00Z")
      .expect(200);

    expect(body).toMatchObject({
      active: 0,
      inactive: 0,
      tokenGenerated: 0,
      closedWon: 0,
      downPayment: 0,
    });
  });

  [
    "",
    "fromDate=2021-01-01T00:00:00Z",
    "toDate=2021-04-15T00:00:00Z",
    "fromDate=2021-01-01T00:00:00Z&toDate=2021-04-15T00:00:00Z",
  ].forEach((queryParams) => {
    it(`accepts "${queryParams}" as date range if the provided date is ISO8601 date and returns the leadsSummary as per the provided query params`, async () => {
      const { body } = await useRequestFactory()
        .withVerifiedUser()
        .get(`/leads/summary?${queryParams}`)
        .expect(200);

      expect(body).toMatchSnapshot();
    });
  });

  ["fromDate=true", "fromDate=156", "toDate=helloworld", "fromDate=null&toDate=undefined"].forEach(
    (queryParams) => {
      it(`throws Bad Request for the date range = "${queryParams}" since the provided date is not ISO8601 date`, async () => {
        const { body } = await useRequestFactory()
          .withVerifiedUser()
          .get(`/leads/summary?${queryParams}`)
          .expect(400);

        expect(body).toMatchSnapshot();
      });
    }
  );
});
