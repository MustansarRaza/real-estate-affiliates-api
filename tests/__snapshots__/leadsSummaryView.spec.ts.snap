// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /leads/summary accepts "" as date range if the provided date is ISO8601 date and returns the leadsSummary as per the provided query params 1`] = `
Object {
  "active": 3,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 1,
  "tokenGenerated": 0,
}
`;

exports[`GET /leads/summary accepts "fromDate=2021-01-01T00:00:00Z" as date range if the provided date is ISO8601 date and returns the leadsSummary as per the provided query params 1`] = `
Object {
  "active": 1,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 3,
  "tokenGenerated": 0,
}
`;

exports[`GET /leads/summary accepts "fromDate=2021-01-01T00:00:00Z&toDate=2021-04-15T00:00:00Z" as date range if the provided date is ISO8601 date and returns the leadsSummary as per the provided query params 1`] = `
Object {
  "active": 1,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 3,
  "tokenGenerated": 0,
}
`;

exports[`GET /leads/summary accepts "toDate=2021-04-15T00:00:00Z" as date range if the provided date is ISO8601 date and returns the leadsSummary as per the provided query params 1`] = `
Object {
  "active": 3,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 1,
  "tokenGenerated": 0,
}
`;

exports[`GET /leads/summary denies access to unauthenticated requests 1`] = `Object {}`;

exports[`GET /leads/summary denies access to unauthenticated requests 2`] = `
Object {
  "cache-control": "private, no-store",
  "connection": "close",
  "strict-transport-security": "max-age=15552000; includeSubDomains",
  "transfer-encoding": "chunked",
  "x-content-type-options": "nosniff",
  "x-dns-prefetch-control": "off",
  "x-download-options": "noopen",
  "x-frame-options": "SAMEORIGIN",
  "x-xss-protection": "1; mode=block",
}
`;

exports[`GET /leads/summary returns leads summary for a verified user 1`] = `
Object {
  "active": 1,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 3,
  "tokenGenerated": 0,
}
`;

exports[`GET /leads/summary returns leads summary for a verified user 2`] = `
Object {
  "cache-control": "private, no-store",
  "connection": "close",
  "content-length": "98",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"62-8SaL6LP4INuUZPpFPUELDSEAzXs\\"",
  "strict-transport-security": "max-age=15552000; includeSubDomains",
  "vary": "Accept-Encoding",
  "x-content-type-options": "nosniff",
  "x-dns-prefetch-control": "off",
  "x-download-options": "noopen",
  "x-frame-options": "SAMEORIGIN",
  "x-xss-protection": "1; mode=block",
}
`;

exports[`GET /leads/summary throws Bad Request for the date range = "fromDate=156" since the provided date is not ISO8601 date 1`] = `
Object {
  "errors": Array [
    Object {
      "location": "query",
      "message": "Not a valid ISO 8601 date",
      "name": "InvalidRequestData",
      "param": "fromDate",
    },
  ],
}
`;

exports[`GET /leads/summary throws Bad Request for the date range = "fromDate=null&toDate=undefined" since the provided date is not ISO8601 date 1`] = `
Object {
  "errors": Array [
    Object {
      "location": "query",
      "message": "Not a valid ISO 8601 date",
      "name": "InvalidRequestData",
      "param": "fromDate",
    },
    Object {
      "location": "query",
      "message": "Not a valid ISO 8601 date",
      "name": "InvalidRequestData",
      "param": "toDate",
    },
  ],
}
`;

exports[`GET /leads/summary throws Bad Request for the date range = "fromDate=true" since the provided date is not ISO8601 date 1`] = `
Object {
  "errors": Array [
    Object {
      "location": "query",
      "message": "Not a valid ISO 8601 date",
      "name": "InvalidRequestData",
      "param": "fromDate",
    },
  ],
}
`;

exports[`GET /leads/summary throws Bad Request for the date range = "toDate=helloworld" since the provided date is not ISO8601 date 1`] = `
Object {
  "errors": Array [
    Object {
      "location": "query",
      "message": "Not a valid ISO 8601 date",
      "name": "InvalidRequestData",
      "param": "toDate",
    },
  ],
}
`;
